spring-cloud-gcp-pubsub/src/main/java/org/springframework/cloud/gcp/pubsub/support/DefaultPublisherFactory.java

```diff
@@ -52,37 +50,92 @@
 	 */
 	private final ConcurrentHashMap<String, Publisher> publishers = new ConcurrentHashMap<>();
 
-	private final ExecutorProvider executorProvider;
+	private ExecutorProvider executorProvider;
+
+	private TransportChannelProvider channelProvider;
+
+	private CredentialsProvider credentialsProvider;
 
-	private final TransportChannelProvider channelProvider;
+	private RetrySettings retrySettings;
 
-	private final CredentialsProvider credentialsProvider;
+	private BatchingSettings batchingSettings;
 
-	public DefaultPublisherFactory(GcpProjectIdProvider projectIdProvider,
-			ExecutorProvider executorProvider,
-			TransportChannelProvider channelProvider,
-			CredentialsProvider credentialsProvider) {
+	/**
+	 * @param projectIdProvider provides the GCP project ID
+	 */
+	public DefaultPublisherFactory(GcpProjectIdProvider projectIdProvider) {
 		Assert.notNull(projectIdProvider, "The project ID provider can't be null.");
-		Assert.notNull(executorProvider, "The executor provider can't be null.");
-		Assert.notNull(channelProvider, "The channel provider can't be null.");
-		Assert.notNull(credentialsProvider, "The credentials provider can't be null.");
 
 		this.projectId = projectIdProvider.getProjectId();
 		Assert.hasText(this.projectId, "The project ID can't be null or empty.");
+	}
+
+	/**
+	 * @param executorProvider provides the executor to be used by the publisher, useful to specify
+	 *                         the number of threads to be used by each executor
+	 */
+	public void setExecutorProvider(ExecutorProvider executorProvider) {
 		this.executorProvider = executorProvider;
```

Comment:
Does it make sense to just create the empty builder in the constructor, and set directly on the builder here?
```java
this.publisherBuilder.setExecutorProvider(executorProvider);
```
This way you can remove all of those instance variables and also simplify the `getPublisher` to a simple `this.publisherBuilder.build()`.

Reply:
I love the idea, but don't think it's feasible because `Subscriber.newBuilder()` takes a `SubscriptionName`, which we don't have at constructor time.

We can't really add it to the constructor because we'll go from 1 factory -> n subscriptions to 1 factory -> 1 subscription, at which point `SubscriberFactory` would just be a proxy for `Subscriber.Builder` (which is also not easily mockable).

Reply:
Yeah, I see, it's a bummer. :(